src/software/new_geom/rectangle.h

```diff
@@ -0,0 +1,112 @@
+#include "software/new_geom/convex_polygon.h"
+
+/**
+ * A rectangle.
+ */
+class Rectangle : public ConvexPolygon
+{
+   public:
+    /**
+     * Creates a new Rectangle from two corners.
+     *
+     *   *--------p1/2      p1/2--------*
+     *   |          |        |          |
+     * p1/2---------*   or   *--------p1/2
+     *
+     * @param point1 One of the rectangle's corners
+     *
+     * @param point2 The corner diagonally-opposite to point1
+     */
+    explicit Rectangle(const Point &point1, const Point &point2);
+
+    /**
+     * Returns the length along the x-axis of the rectangle
+     *
+     * @return The length along the x-axis of the rectangle
+     */
+    double xLength() const;
+
+    /**
+     * Returns the length along the y-axis of the rectangle
+     *
+     * @return The length along the y-axis of the rectangle
+     */
+    double yLength() const;
+
+    /**
+     * Returns the area of the rectangle
+     *
+     * @return The area of the rectangle
+     */
+    double area() const override;
+
+    /**
+     * Returns the centre of the rectangle
+     *
+     * @return The centre of the rectangle
+     */
+    Point centre() const;
+
+    /**
+     * Returns the <+x,+y> corner of the rectangle
+     *
+     * @return The <+x,+y> corner of the rectangle
+     */
+    Point posXPosYCorner() const;
+
+    /**
+     * Returns the <-x,+y> corner of the rectangle
+     *
+     * @return The <-x,+y> corner of the rectangle
+     */
+    Point negXPosYCorner() const;
+
+    /**
+     * Returns the <-x,-y> corner of the rectangle
+     *
+     * @return The <-x,-y> corner of the rectangle
+     */
+    Point negXNegYCorner() const;
+
+    /**
+     * Returns the <+x,-y> corner of the rectangle
+     *
+     * @return The <+x,-y> corner of the rectangle
+     */
+    Point posXNegYCorner() const;
+
+    /**
+     * Determines whether the given Point is contained within this Rectangle.
+     *
+     * @return whether the Point p is contained within this Rectangle.
+     */
+    bool contains(const Point &p) const override;
+
+    /**
+     * Returns a digit corresponding to a specific corner for pos%4. (0 for <-x,-y>, 1
+     * for <-x,+y>, 2 for <+x,+y> and 3 for <+x,-y>)
+     *
+     * @param pos The position of the corner wanted
+     *
+     * @return Point A digit corresponding to position
+     */
+    Point operator[](unsigned int pos) const;
```

Comment:
Do we actually use this function in our code? IMO I don't think we should have it since it exposes the internal representation of the `Rectangle`.

Reply:
Removed in latest update